### PR TITLE
[board-server] #2889 Allow board server to use SQLite and run self-contained in a docker container

### LINKED ISSUE #2889: [board-server] Allow Board Server to run in a container as a self-contained deployment
The Board Server is an incredibly exciting development for the share-ability of boards, but currently the means of deploying it is very GCP-centric, and the storage depends on Firestore. I would like to:

- localise the storage out of firestore - maybe into SQLite or similar.
- write a dockerfile that builds a completely self-contained board server container.

### PR DESCRIPTION
**Pull request by wfaithfull**
Fixes #2889 

I think this is approaching time to merge as I've overcome the main pain points. 

There is a bunch more I want to do, but I would prefer to get these changes reviewed before embarking on anything else. Follow up plans:

- Database migration handling
- A github actions workflow that builds the container and publishes an attested build to GHCR.io 
- Improve the dockerfile in a number of ways 
  - just have the prod build for image size
  - define a volume for the sqlite db

Highlights:

### The board server now supports SQLite as a storage backend

You can control this behaviour (currently) with environment variables. We still default to firestore right now.

```
export STORAGE_BACKEND=sqlite
export SQLITE_DB_PATH=foo.db
```

The firestore and sqlite storage providers are separated out behind an interface, find them in `src/server/storage-providers`.

### Error handling

The board server now responds 500 Internal Server Error on unhandled exceptions.

### Board API integration tests

There's now a (admittedly, only happy path) integration test suite that puts the board API through it's paces in `tests/integration/board-api.test.ts`. Run this locally against SQLite with a new target:

```
npm run test:integration
``` 

### A Dockerfile for the board server

Now that we have the option to run with SQLite, we can run the board server in a fully self contained docker image. 

To build the container:

```
cd packages/board-server # if not already
docker build --build-context breadboard=../../ -t board-server .
```

To run the container:

```
docker run -d -p 3000:3000 board-server
```

### CODE REVIEW

**Comment by wfaithfull** [comment]
@dglazkov @aomarks would appreciate your 👀 on this. I've added a minor bump changeset for the moment, as in the course of doing this I have started handling errors and returning 500 responses. If anyone was depending on that I feel sorry for them, but nonetheless..

It turned out to be a much bigger task than I first thought, so I'm expecting a few comments 😄

**Comment by timswanson-google** [comment]
Taking a look now. I'll probably be able to have a review done by North American Tuesday.

File: packages/board-server/tests/integration/board-api.test.ts [hunk 1]
```diff
@@ -0,0 +1,489 @@
+import { suite, test } from 'node:test';
```
Comment (dglazkov) [inline packages/board-server/tests/integration/board-api.test.ts:1]: Wow, this is great! Thank you for writing these.

File: packages/board-server/tsconfig.json [hunk 2]
```diff
@@ -20,7 +20,8 @@
     "useUnknownInCatchVariables": true,
     "experimentalDecorators": true,
     "useDefineForClassFields": false,
-    "resolveJsonModule": true
+    "resolveJsonModule": true,
+    "esModuleInterop": true
```
Comment (dglazkov) [inline packages/board-server/tsconfig.json:24]: Curious, why did you need that?

File: packages/board-server/src/server/types.ts [hunk 3]
```diff
@@ -111,8 +112,22 @@ export type RunBoardStateStore = {
     ticket: string
   ): Promise<ReanimationState | undefined>;
   saveReanimationState(user: string, state: ReanimationState): Promise<string>;
+  getBoardServerCorsConfig(): Promise<BoardServerCorsConfig | undefined>
```
Comment (dglazkov) [inline packages/board-server/src/server/types.ts:115]: I would do something different here. The `RunBoardStateStore` is a subset of the larger `BoardServerStore` type that needs to be defined. Then, `BoardServerStore` extends it. The `RunBoardStateStore` is specifically about storing the reanimation state. But this is something that can be done as a follow-up patch.

File: packages/board-server/tests/run-board.ts [hunk 4]
```diff
@@ -139,6 +141,30 @@ const runStateStore = {
   async saveReanimationState(user: string, state: any) {
     return JSON.stringify(state);
   },
+  getBoardServerCorsConfig: function (): Promise<BoardServerCorsConfig | undefined> {
```
Comment (dglazkov) [inline packages/board-server/tests/run-board.ts:144]: See above. If we have `BoardServerStore`, then we don't need to reimplement all of these.

**Review by dglazkov — COMMENTED** [review]
Amazing work! Will let @timswanson-google have the final look/say.

File: .gitignore [hunk 5]
```diff
@@ -15,6 +15,12 @@ pids
 *.seed
 *.pid.lock
 
+# SQLite board server local db
+*.db
+
+# output from markdown task
+all_markdown.md
```
[the remark this answers is not in the document]
Comment (wfaithfull) [reply .gitignore:22]: This was something I missed out from the previous PR that introduced that task, so I thought since I was making changes to the gitignore anyway I would add this at the same time.

File: packages/board-server/src/server/common.ts [hunk 6]
```diff
@@ -37,29 +54,22 @@ const CONTENT_TYPE = new Map([
...
-
-/**
- * Serve a static file
- */
 export const serveFile = async (
   res: ServerResponse,
   path: string,
   transformer?: (contents: string) => Promise<string>
 ) => {
+  console.log(`Attempting to serve file: ${path}`);
```
[the remark this answers is not in the document]
Comment (wfaithfull) [reply packages/board-server/src/server/common.ts:62]: Thank you, yes - leftover from my attempts to debug the pathing

File: packages/board-server/src/server/common.ts [hunk 7]
```diff
@@ -3,16 +3,33 @@ import type { IncomingMessage, ServerResponse } from "http";
...
 
 import type { ServerConfig } from "./config.js";
 import { notFound, serverError } from "./errors.js";
 import type { PageMetadata } from "./types.js";
 import { createReadStream } from "fs";
 
 const MODULE_PATH = dirname(fileURLToPath(import.meta.url));
-const ROOT_PATH = resolve(MODULE_PATH, "../..");
+
+function findRootPath() {
```
[the remark this answers is not in the document]
Comment (wfaithfull) [reply packages/board-server/src/server/common.ts:15]: So I suspect this is a function of my inexperience with JS/TS/Vite. I found that when running from the tests, the root path that was getting resolved was always one level wrong, and this meant that the root path included `/dist` on the end, but so did the `client_path`.

So in a sense, I've bailed out with this more sophisticated pathing approach. If somebody else has more insight I would welcome it. Perhaps @jamesjacko or @Mearman might have a better idea why I was facing this issue?

File: packages/board-server/tsconfig.json (hunk 2 above)
Comment (wfaithfull) [reply packages/board-server/tsconfig.json:24]: Again, my experience of TS/JS is fairly limited so perhaps there is a better solution, however the `node:sqlite` library which `better-sqlite3` uses uses commonjs - I couldn't find a sqlite library that didn't require some kind of commonjs support. Open to suggestions here.

File: packages/board-server/src/server/types.ts (hunk 3 above)
Comment (wfaithfull) [reply packages/board-server/src/server/types.ts:115]: Yeah, I agree - I actually started down this path, having a `BoardServerStore` and `RunBoardStateStore` but I ran into difficulties with union types and then forgot to address it again. I'll have a play with it today and see if I can get it to work, but if not, I'll do this in a follow up.

File: packages/board-server/src/server/common.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Mearman) [reply packages/board-server/src/server/common.ts:15]: I have a feeling that the issue may be because the value is declared as a `const`. If you compute the "`base path`" at the apex of the execution and pass it down via the `makeRouter`, it might resolve it.

[the remark this answers is not in the document]
Comment (wfaithfull) [reply packages/board-server/src/server/common.ts:15]: Ooh, thank you for the suggestion @Mearman, let me have a play with that idea..

[the remark this answers is not in the document]
Comment (wfaithfull) [reply packages/board-server/src/server/common.ts]: I've incorporated your suggestion @Mearman in [f4d514c](https://github.com/breadboard-ai/breadboard/pull/2933/commits/f4d514c31501d21fd5ca79a499609a50f40d08db). Thank you for that, I think I finally understand the problem. When starting the server from a test context, the module path is `dist/tests/integration`, so I need to ascend three levels rather than two. I have added a parameter in the server start to override the root path from tests - putting this in the server config which is already passed around, this leaves the main logic in common.ts untouched.

File: packages/board-server/src/server/store.ts [hunk 8]
```diff
@@ -28,10 +18,44 @@ export type OperationResult =
...
+  if (provider === "sqlite" || provider === "firestore") {
+    return provider;
+  }
+  
+  return "firestore";
+}
+
 export const getStore = () => {
-  return new Store("board-server");
+  return providers[getProvider()]
```
Comment (timswanson-google) [inline packages/board-server/src/server/store.ts:37]: It feels like we could combine these two functions:

```
export const getStore = () => {
  return providers[process.env['STORAGE_BACKEND']] ?? providers["firestore"]!;
}
```

(or even just an `if/else` block, which is basically the same thing with less indirection)

File: packages/board-server/src/server/storage-providers/firestore.ts
Comment (timswanson-google) [inline packages/board-server/src/server/storage-providers/firestore.ts:1]: (skipping review on this file since I'm assuming it's just a move of what used to be in store.ts. let me know if that's not the case.)

File: packages/board-server/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,10 @@
+FROM node:20
+COPY --from=breadboard . /breadboard/
+# For the time being, we will assume container deployments are exclusively sqlite.
+ENV STORAGE_BACKEND "sqlite"
+ENV GOOGLE_CLOUD_PROJECT "n/a"
```
Comment (timswanson-google) [inline packages/board-server/Dockerfile:5]: I think we need to set ALLOWED_ORIGINS here for the board server to accept requests. (Similar to what we do in `app.yaml`. Since each deployment will need a different value, I've just set it to empty string so that the person deploying has a placeholder to fill in.

By default it will reject all connections.

File: packages/board-server/src/server/storage-providers/sqlite.ts [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+  
+      const state = JSON.parse(row.state);
+      if (!state.states) {
+        return undefined;
+      }
+  
+      return state;
+    }
+  
+    async getBoardServerCorsConfig(): Promise<BoardServerCorsConfig | undefined> {
```
Comment (timswanson-google) [inline packages/board-server/src/server/storage-providers/sqlite.ts:93]: FYI this table is unused and should be deleted. I'll wait to remove until this PR is in, to avoid merge conflicts, but if you wanted this just to returned undefined (and to remove all references to the configuration table) it wouldn't have any impact on server behavior.

File: packages/board-server/tests/integration/board-api.test.ts [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+process.on('unhandledRejection', async (reason) => {
+  console.error('Unhandled Rejection:', reason);
+  if (serverInstance) {
+    await stopServer(serverInstance.server);
+  }
+  process.exit(1);
+});
+
+function readBoard(name: string) {
+      // Get the directory name
```
Comment (timswanson-google) [inline packages/board-server/tests/integration/board-api.test.ts:51]: indentation (running the formatter should probably pick this up)

File: packages/board-server/tests/integration/board-api.test.ts [hunk 12]
```diff
@@ -0,0 +1,494 @@
...
+    return next;
+  }
+  if (type === "output" || type === "end") {
+    return undefined;
+  }
+  throw new Error(`Unexpected state type: ${type}`);
+};
+
+suite('Board API Integration tests', async () => {
+// Tests
```
Comment (timswanson-google) [inline packages/board-server/tests/integration/board-api.test.ts:236]: is this a stray comment?

File: packages/board-server/tests/integration/board-api.test.ts [hunk 13]
```diff
@@ -0,0 +1,494 @@
...
+      path: `/boards/@${account.account}/simple.json?API_KEY=${account.api_key}`,
+      method: 'POST',
+      body: readBoard('simple.bgl.json'),
+      headers: { 'Content-Type': 'application/json' }
+    });
+    assert.strictEqual(statusCode, 200);
+    assert.deepStrictEqual(JSON.parse(body), { "created" : `@${account.account}/simple.json` });
+  });
+
+  test('GET /boards/@test/simple.json should get a board', { concurrency: false }, async () => {
```
Comment (timswanson-google) [inline packages/board-server/tests/integration/board-api.test.ts:257]: Do these tests need to run in order? Does this test's success depend on the previous on having been run?

File: packages/board-server/tests/integration/board-api.test.ts (hunk 1 above)
Comment (timswanson-google) [reply packages/board-server/tests/integration/board-api.test.ts:1]: Seriously. The hardest part of tests is writing them in the first place. It's so much easier to modify tests that already exist than to create them from scratch. Thank you.

File: packages/board-server/src/server/storage-providers/firestore.ts
Comment (wfaithfull) [reply packages/board-server/src/server/storage-providers/firestore.ts:1]: Correct, it's a lift-and-shift.

File: packages/board-server/tests/integration/board-api.test.ts (hunk 13 above)
Comment (wfaithfull) [reply packages/board-server/tests/integration/board-api.test.ts:257]: They do, I'm afraid. Not an ideal state of affairs but it brought the horizon for this test suite a lot closer. I think I can bring a follow up patch to resolve this now that I have a much better understanding of the API.

File: packages/board-server/tests/integration/board-api.test.ts (hunk 12 above)
Comment (wfaithfull) [reply packages/board-server/tests/integration/board-api.test.ts:236]: Yes, thank you for spotting that!

File: packages/board-server/tests/integration/board-api.test.ts (hunk 11 above)
Comment (wfaithfull) [reply packages/board-server/tests/integration/board-api.test.ts:51]: Thank you!

File: packages/board-server/src/server/boards/index.ts [hunk 14]
```diff
@@ -105,11 +105,11 @@ export const serveBoardsAPI = async (
     }
     case "app": {
       // Serve the index.html file for the app.
-      serveIndex(serverConfig.viteDevServer, res, getMetadata(parsed));
+      serveIndex(serverConfig, serverConfig.viteDevServer, res, getMetadata(parsed));
```
[the remark this answers is not in the document]
Comment (wfaithfull) [reply packages/board-server/src/server/boards/index.ts:108]: Yes, good idea.

File: packages/board-server/src/server/store.ts (hunk 8 above)
Comment (wfaithfull) [reply packages/board-server/src/server/store.ts:37]: Yes, that's a nice suggestion. I'll take a stab at it.

File: packages/board-server/Dockerfile (hunk 9 above)
Comment (wfaithfull) [reply packages/board-server/Dockerfile:5]: Ah, excellent spot. Ok, I will investigate.

File: packages/board-server/src/server/storage-providers/sqlite.ts (hunk 10 above)
Comment (wfaithfull) [reply packages/board-server/src/server/storage-providers/sqlite.ts:93]: I think I'll leave it in for now as it makes no difference to the upcoming removal, and then that follow up diff will make more sense.

File: packages/board-server/tests/integration/board-api.test.ts (hunk 13 above)
Comment (wfaithfull) [reply packages/board-server/tests/integration/board-api.test.ts]: To be clear - setting concurrency false and wrapping them in a suite ensures they run in order, so I'm not leaving it up to chance or interpretation.

**Comment by timswanson-google** [comment]
Merging. If there's anything else we want to do we'll get it in a followup. Thanks Will!
